src/r1cs/verifier.rs

```diff
@@ -94,8 +86,24 @@ impl<'a, 'b> ConstraintSystem for VerifierCS<'a, 'b> {
         self.constraints.push(lc);
     }
 
-    fn challenge_scalar(&mut self, label: &'static [u8]) -> Scalar {
-        self.transcript.challenge_scalar(label)
+    fn randomized_constraints<F>(
+        &mut self,
+        label: &'static [u8],
+        callback: F,
+    ) -> Result<(), R1CSError>
+    where
+        F: 'static + Fn(&mut Self, Scalar) -> Result<(), R1CSError>,
+    {
+        match self.deferred_constraints {
+            Some(ref mut list) => {
+                list.push((label, Box::new(callback)));
```

Comment:
I'm not sure I understand what's happening here... if
```
callback: F,
F: 'static + Fn(&mut Self, Scalar) -> Result<(), R1CSError>,
```
then why do we need to box `callback`? Isn't it already `'static` and will therefore outlive the `Verifier`?  Alternatively, if we are going to box it, does it need to be `'static`?

Reply:
Hmm. Good question. I think this is an artifact of me-not-knowing-Rust-well tinkering in #196 and ending up with something that finally compiles.

I guess i'll try to see if we can keep `'static` and avoid boxing?

Reply:
Seems like we need to do `Box<'static+Fn...>`. Here's why:

1. Closures are all of different (anonymous) type as they are created by distinct gadgets, but we need to store them in a `Vec`, so they have to be wrapped in a single type. Hence, boxing. In other words, `Fn` is a trait, not a type.
2. Closure is owned by box, which erases lexical lifetime and therefore requires `'static`. This, in turn, implies that variables can be only captured by move into closure, not by reference (unless the references are 'static themselves).

Reply:
Cool, requiring a trait object to stick them in a `Vec` makes sense.